subprojects/kotlin-dsl-provider-plugins/src/main/kotlin/org/gradle/kotlin/dsl/provider/plugins/precompiled/PrecompiledScriptPlugin.kt

```diff
@@ -139,19 +139,122 @@ private
 fun packageNameOf(code: String): String? =
     KotlinLexer().run {
         start(code)
+
+        // Based on: https://kotlinlang.org/docs/reference/grammar.html#kotlinFile
+        //
+        // The only elements Kotlin script grammar allows before the package statement
+        // are the shebang line and file annotations.
+        //
+        // If neither of them is there, then we know that the package statement is
+        // right at the beginning of the script, if present at all, so we can do a
+        // very simple and cheap search for it.
+        //
+        // Otherwise, we must do a bit more involved searching.
+
         skipWhiteSpaceAndComments()
-        when (tokenType) {
-            KtTokens.PACKAGE_KEYWORD -> {
-                advance()
-                skipWhiteSpaceAndComments()
-                parseQualifiedName()
-            }
+        if (hasShebangLine() || hasFileAnnotations()) {
+            return extensiveSearchForPackageName()
+        } else {
+            return simpleSearchForPackageName()
+        }
+    }
+
+
+private
+fun KotlinLexer.hasShebangLine(): Boolean {
+
+    if (tokenType != KtTokens.HASH) {
+        return false
+    }
+
+    val mark = currentPosition
+    advance()
+
+    val hasShebangLine = tokenType == KtTokens.EXCL
+
+    restore(mark)
+    return hasShebangLine
+}
+
+
+private
+fun KotlinLexer.hasFileAnnotations(): Boolean {
+
+    if (tokenType != KtTokens.AT) {
+        return false
+    }
+
+    val mark = currentPosition
+    advance()
 
-            else -> null
+    val hasFileAnnotation = tokenType == KtTokens.IDENTIFIER && tokenText == "file"
+
+    restore(mark)
+    return hasFileAnnotation
+}
+
+
+private
+fun KotlinLexer.simpleSearchForPackageName(): String? =
+    when (tokenType) {
+        KtTokens.PACKAGE_KEYWORD -> {
+            advance()
+            skipWhiteSpaceAndComments()
+            parseQualifiedName()
         }
+
+        else -> null
     }
 
 
+private
+fun KotlinLexer.extensiveSearchForPackageName(): String? {
+    while (hasFurtherTokens() && !packageDefinitionFound() && !tooLateForPackageStatement()) {
+        advance()
+    }
+
+    if (packageDefinitionFound()) {
+        advance()
+        skipWhiteSpaceAndComments()
+        return parseQualifiedName()
+    }
+
+    return null
+}
+
+
+private
+fun KotlinLexer.hasFurtherTokens() = tokenType != null
+
+
+private
+fun KotlinLexer.packageDefinitionFound() = tokenType == KtTokens.PACKAGE_KEYWORD
+
+
+private
+fun KotlinLexer.tooLateForPackageStatement(): Boolean {
+    // based on https://kotlinlang.org/docs/reference/grammar.html#script
+
+    if (tokenType == KtTokens.IDENTIFIER && tokenText in setOf("import", "buildscript", "plugins", "pluginManagement", "initscript")) {
+        return true
+    }
+
+    return tokenType in setOf(
```

Comment:
:lipstick: Maybe extract the `Set` out of this function as a `private val` so it doesn't need to be created for each lexed script.